Include/internal/pycore_dict.h

```diff
@@ -264,6 +264,37 @@ _PyDictValues_AddToInsertionOrder(PyDictValues *values, Py_ssize_t ix)
     *size_ptr = size;
 }
 
+#ifdef Py_GIL_DISABLED
+
+static inline bool
+_PyDictOrValues_TryStoreValue(PyObject *obj, Py_ssize_t index, PyObject *value)
+{
+    bool success;
+    Py_BEGIN_CRITICAL_SECTION(obj);
+
+    PyDictOrValues *dorv_ptr = _PyObject_DictOrValuesPointer(obj);
+    if (!_PyDictOrValues_IsValues(*dorv_ptr)) {
+        success = false;
+        goto exit;
+    }
+
+    PyDictValues *values = _PyDictOrValues_GetValues(*dorv_ptr);
```

Comment:
I think it'd be better if the check and the retrieval of the values were combined instead of having multiple loads of the values array.

Reply:
With fixing the atomic loads the compiler should be able to optimize this down to a single read.  This is basically what the pattern typically was before this change and adding `_PyDictOrValues_TryGetValues` which I don't want to use here because we're already locking.